### PR TITLE
Add types to JoinTerms

### PR DESCRIPTION
**Pull request by daxida**
We need types for writing `WriteJoin` commands in C.

I quickly hacked something in `parser.rs` and tested it. Picture below.

The code itself for adding the types is not ideal and I would like to have your opinion before proceeding. 

image

Also:
- (FIXED) `var` should be `term` in `write_join.c.jinja`
- Changed `name` to `ident` in JoinTerm for consistency
- Added `var_type` to JoinTerm (now is_variable became useless since the info is already in the Option var_type, but may be worth keeping for readability)

Edit:
For the moment it can be tested with this handmade clash (but ideally this should be standard and accessible through `gen --debug` or sth):
```
cargo run fetch 90435e82d1d5e3fe5f9d3dd813770f0d5a7d2
cargo run next 90435e82d1d5e3fe5f9d3dd813770f0d5a7d2
cargo run gen c
```

### CODE REVIEW

File: src/stub/parser.rs [hunk 1]
```diff
@@ -45,6 +45,27 @@ impl<'a> Parser<'a> {
...
+            if let Cmd::WriteJoin { join_terms, output_comment: _ } = cmd {
+                for term in join_terms.iter_mut() {
+                    for (ident, var_type) in &read_pairings {
+                        if term.ident == *ident {
+                            term.var_type = Some(var_type.clone());
+                        }
+                    }
+                }
+            }
+        }
```
Comment (Andriamanitra) [inline src/stub/parser.rs]: This probably makes no difference but at least we can get rid of one of the loops.
I also added a `#[derive(Copy)]` for the `VarType` enum so there's no need to explicitly .clone() `var_type` (copying enums without complicated data is cheap).

```suggestion
        let mut read_pairings = std::collections::BTreeMap::new();
        for read_cmd in &stub.commands {
            if let Cmd::Read(var_cmds) = read_cmd {
                for var_cmd in var_cmds {
                    read_pairings.insert(var_cmd.ident.clone(), var_cmd.var_type);
                }
            }
        }        

        for cmd in &mut stub.commands {
            if let Cmd::WriteJoin { join_terms, output_comment: _ } = cmd {
                for term in join_terms.iter_mut() {
                    if let Some(var_type) = read_pairings.get(&term.ident) {
                        term.var_type = Some(*var_type);
                    }
                }
            }
        }
```

Comment (daxida) [reply src/stub/parser.rs]: Sure, I added it in [57ef078](https://github.com/ellnix/clash/pull/30/commits/57ef0788d99357b5e97b572e687d1723d7d5020e)

File: config/stub_templates/c/write_join.c.jinja [hunk 2]
```diff
@@ -1,35 +1,17 @@
...
-  {%- set_global joined_terms = joined_terms | concat(with=ref) -%}
-{%- endfor -%}
-{%- set fmt_joined_terms = joined_terms | join(sep=" ") -%}
-
 {%- set_global types = [] -%}
-{%- set_global var_refs = [] -%}
-{%- for var in terms -%}
+{%- set_global term_refs = [] -%}
+{%- for term in terms -%}
   {# var_type is None when the term is a literal #}
```
Comment (ellnix) [inline config/stub_templates/c/write_join.c.jinja:4]: Maybe we should remove the `term.is_variable` bool in the future.

File: src/main.rs [hunk 3]
```diff
@@ -487,13 +488,44 @@ impl App {
 
     fn generate_stub(&self, args: &ArgMatches) -> Result<()> {
         let config = self.build_stub_config(args)?;
-        let handle = self
-            .current_handle()
-            .expect("You must have a current clash to generate stubs. Please use clash next");
-        let clash = self.read_clash(&handle)?;
-        let stub_generator = clash.stub_generator().expect("Clash provides no input stub generator");
+        let reference_generator = r##"read anInt:int
```
Comment (ellnix) [inline src/main.rs:491]: This should maybe be a constant?

**Review by ellnix — APPROVED** [review]

**Pull request merged**